### PR TITLE
nerdfonts: expand capabilities

### PR DESCRIPTION
**Pull request by eadwu**
Allow patching of fonts from derivation outputs
Fix output path of *.otf files

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/nerdfonts/default.nix [hunk 1]
```diff
@@ -1,26 +1,50 @@
...
   src = fetchFromGitHub {
     owner = "ryanoasis";
     repo = "nerd-fonts";
     rev = version;
     sha256 = "1la79y16k9rwcl2zsxk73c0kgdms2ma43kpjfqnq5jlbfdj0niwg";
   };
-  dontPatchShebangs = true;
-  buildInputs = [ which ];
+
+  buildInputs = [ which fontforge ];
```
Comment (rnhmjoj) [inline pkgs/data/fonts/nerdfonts/default.nix]:
```suggestion
  nativeBuildInputs = [ which fontforge ];
```

File: pkgs/data/fonts/nerdfonts/default.nix [hunk 2]
```diff
@@ -1,26 +1,50 @@
...
+      find ${drv.out}/share/fonts -name '*.otf' \
+        -exec $ffPatcher --outputdir $out/share/fonts/opentype --extension otf "{}" \;
+      find ${drv.out}/share/fonts -name '*.ttf' \
+        -exec $ffPatcher --outputdir $out/share/fonts/truetype --extension ttf "{}" \;
+    '') patchFontDrvs}
+
+    ${lib.optionalString (fonts != null) ''
+      ./install.sh ${lib.concatStringsSep " " fonts}
+
+      find $out/share/fonts -name '*.otf' -exec mv "{}" $out/share/fonts/opentype \;
```
Comment (rnhmjoj) [inline pkgs/data/fonts/nerdfonts/default.nix]: Is it not possible to do `install -D *.otf -t $out/share/fonts/opentype`?
This would also shave off the `mkdir -p`

File: pkgs/data/fonts/nerdfonts/default.nix [hunk 3]
```diff
@@ -1,26 +1,50 @@
-{ stdenv, fetchFromGitHub, which, withFont ? "" }:
+{ lib, stdenv, fetchFromGitHub
+, which, fonts ? []
+, patchFontDrvs ? [], patchFlags ? [ "mono" "windows" "complete" "careful" ], fontforge ? null }:
```
Comment (rnhmjoj) [inline pkgs/data/fonts/nerdfonts/default.nix]: Is this a list of fonts that will be patched and added to the output of nerdfonts?
I think this warrants a section in the [packages](https://nixos.org/nixpkgs/manual/#chap-packages) chapter of the manual.

File: pkgs/data/fonts/nerdfonts/default.nix [hunk 4]
```diff
@@ -1,26 +1,50 @@
...
+
+    ${lib.optionalString (fonts != null) ''
+      ./install.sh ${lib.concatStringsSep " " fonts}
+
+      find $out/share/fonts -name '*.otf' -exec mv "{}" $out/share/fonts/opentype \;
+    ''}
   '';
 
-  meta = with stdenv.lib; {
+  dontPatchShebangs = true;
```
Comment (rnhmjoj) [inline pkgs/data/fonts/nerdfonts/default.nix]: Why is this needed? I don't see any other font package with this.

**Review by rnhmjoj — CHANGES_REQUESTED** [review]

File: pkgs/data/fonts/nerdfonts/default.nix (hunk 4 above)
Comment (eadwu) [reply pkgs/data/fonts/nerdfonts/default.nix]: I'm not sure, it was in the original expression so I just left it, I assume it had to do with the `install.sh` but it isn't installed so it shouldn't be needed.

File: pkgs/data/fonts/nerdfonts/default.nix (hunk 3 above)
Comment (eadwu) [reply pkgs/data/fonts/nerdfonts/default.nix]: `fonts` has the same purpose of `withFont` but `patchFontDrvs` is meant for patching the fonts through packages provided through Nix, i.e. `pkgs.anonymousPro`. 

The reason the PR is a draft is also because of this, I'm not sure how packages that reference other dependencies should be treated.

File: pkgs/data/fonts/nerdfonts/default.nix (hunk 2 above)
Comment (eadwu) [reply pkgs/data/fonts/nerdfonts/default.nix]: Done.

**Comment by michaelpj** [comment]
Perhaps I'm missing something, but *why* should the nerdfonts derivation be responsible for patching *other* fonts? It seems very indirect to get a font from somewhere else and then pass it through the derivation for an unrelated font set to patch it. Perhaps you want something like a `fontforge` patching setup hook that you can add to derivations that need it?

File: pkgs/data/fonts/nerdfonts/default.nix (hunk 2 above)
Comment (eadwu) [reply pkgs/data/fonts/nerdfonts/default.nix]: Errors out if no `*.otf` fonts exists, reverted.

**Comment by eadwu** [comment]
The patching is quite a tedious process, and the resulting fonts are considered to be Nerd Fonts as well. So if were to create a Nerd variant of a certain font, it would take a while to patch all the fonts. 

The nerdfonts derivation is just a clustering of Nerd fonts, so I just added it into it. I needed it to patch a font to use while ricing, but the patcher is pretty bad still (tried on Comfortaa) and it wasn't ideal.

I guess there isn't a reason why this derivation should be responsible for patching other fonts. It could be a hook, but we don't have derivations of all the individual nerdfonts in nixpkgs yet, so I don't see the reason for just leaving a hook hanging around.

**Comment by michaelpj** [comment]
> The patching is quite a tedious process, and the resulting fonts are considered to be Nerd Fonts as well.

What makes them "Nerd Fonts"? The patcher you've got there seems to just depend on `fontforge`, i.e. it's not obviously anything to do with Nerd Fonts at all (and you don't even patch the downloaded Nerd Fonts, as far as I can see!).

At the moment this derivation does what it says on the tin: it packages the Nerd Fonts. If we want to also patch some other fonts and bundle up some fonts together, we can do that in a separate derivation and compose them together - that's nice with Nix and more discoverable to boot.

**Comment by eadwu** [comment]
The patcher is the same patcher they use to generate their fonts.

Anyway, this is probably better off as a hook if someone needs it.
